runtime/ts/keymgmt/webcrypto.ts

```diff
@@ -0,0 +1,276 @@
+        return Promise.resolve("");
+    }
+}
+
+class WebCryptoDeviceKey extends WebCryptoStorableKey<CryptoKeyPair> implements DeviceKey {
+
+    algorithm(): string {
+        return this.publicKey().algorithm();
+    }
+
+    constructor(key: CryptoKeyPair) {
+        super(key);
+    }
+
+    privateKey(): PrivateKey {
+        return new WebCryptoPrivateKey(this.key.privateKey);
+    }
+
+    publicKey(): PublicKey {
+        return new WebCryptoPublicKey(this.key.publicKey);
+    }
+
+    fingerprint(): PromiseLike<string> {
+        return this.publicKey().fingerprint();
+    }
+}
+
+export class WebCryptoKeyGenerator implements KeyGenerator {
+    generateWrappedStorageKey(deviceKey: DeviceKey): PromiseLike<WrappedKey> {
+        const generatedKey: PromiseLike<CryptoKey> = crypto.subtle.generateKey({name: 'AES-GCM', length: 256},
+            true, ["encrypt", "decrypt", "wrapKey", "unwrapKey"]);
+        return generatedKey.then(key => new WebCryptoSessionKey(key))
+            .then(skey => skey.disposeToWrappedKeyUsing(deviceKey.publicKey()));
+    }
+
+    static getInstance() {
+        return new WebCryptoKeyGenerator();
+    }
+
+    generateAndStoreRecoveryKey(): PromiseLike<RecoveryKey> {
+        return undefined;
+    }
+
+    generateDeviceKey(): PromiseLike<DeviceKey> {
+        const generatedKey: PromiseLike<CryptoKeyPair> = crypto.subtle.generateKey(
+            {
+                hash: {name: "SHA-512"},
+                name: 'RSA-OAEP',
+                modulusLength: 2048,
+                publicExponent: new Uint8Array([0x01, 0x00, 0x01]),
+            },
+            false, ["encrypt", "decrypt", "wrapKey", "unwrapKey"]);
+        return generatedKey.then(key => new WebCryptoDeviceKey(key));
+    }
+
+    importKey(pemKey: string): PromiseLike<PublicKey> {
+        const key = rs.KEYUTIL.getKey(pemKey);
+        const jwk = rs.KEYUTIL.getJWKFromKey(key);
+
+        return crypto.subtle.importKey("jwk",
+            jwk as JsonWebKey,
+            {
+                name: "RSA-OAEP",
+                hash: {name: "SHA-1"}
+            }, true, ["encrypt", "wrapKey"]).then(ikey => new WebCryptoPublicKey(ikey));
+    }
+}
+
+export class WebCryptoMemoryKeyStorage implements KeyStorage {
+    storageMap: Map<string, Key>;
+
+    constructor() {
+        this.storageMap = new Map();
+    }
+
+    find(keyFingerPrint: string): PromiseLike<Key> {
+        return Promise.resolve(this.storageMap.get(keyFingerPrint));
+    }
+
+    async write(key: Key): Promise<string> {
+        const id = await key.fingerprint();
+        this.storageMap.set(id, key);
+        return Promise.resolve(key.fingerprint());
+    }
+
+    static getInstance() {
+        return new WebCryptoMemoryKeyStorage();
+    }
+}
+
+export class WebCryptoKeyIndexedDBStorage implements KeyStorage {
+
+    async runOnStore(fn: (store: ObjectStore<{}, IDBValidKey>) => PromiseLike<IDBValidKey>) {
+        const db = await idb.open('ArcsKeyManagement', 1,
```

Comment:
Probably worth `const` defined somewhere for `ArcsKeyManagement` and `ArcsKeyManagementStore` strings.